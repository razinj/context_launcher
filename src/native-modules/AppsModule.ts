import { NativeModules } from 'react-native'

export interface AppsModuleInterface {
  getConstants(): AppsModuleConstants
  launchApplication(packageName: string): void
  showApplicationDetails(packageName: string): void
  requestApplicationUninstall(packageName: string): void
  getApplications(): Promise<string>
}

type AppsModuleConstants = {
  appVersion: string
  buildNumber: string
  packageName: string
}

const { AppsModule } = NativeModules

export default AppsModule as AppsModuleInterface
