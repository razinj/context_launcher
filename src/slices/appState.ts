import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RenderedIn } from '../models/rendered-in'
import { RootState } from '../store'

interface AppsSearchState {
  query?: string
  result: AppDetails[]
}

export interface AppState {
  displayAllApps: boolean
  displaySettings: boolean
  menuAppDetails?: AppDetails
  displayAppMenu: boolean
  displaySortableFavoriteApps: boolean
  displaySortablePinnedApps: boolean
  displaySortableTemporaryPinnedApps: boolean
  search: AppsSearchState
}

const initialState: AppState = {
  displayAllApps: false,
  displaySettings: false,
  menuAppDetails: undefined,
  displayAppMenu: false,
  displaySortableFavoriteApps: false,
  displaySortablePinnedApps: false,
  displaySortableTemporaryPinnedApps: false,
  search: {
    query: undefined,
    result: [],
  },
}

export const appStateSlice = createSlice({
  name: 'appState',
  initialState,
  reducers: {
    // GlobalAppState
    setDisplayAllApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAllApps = payload
    },
    setDisplaySettings: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySettings = payload
    },
    setMenuAppDetails: (state: AppState, { payload }: PayloadAction<AppDetails | undefined>) => {
      state.menuAppDetails = payload
    },
    setDisplayAppMenu: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displayAppMenu = payload
    },
    setDisplaySortableFavoriteApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableFavoriteApps = payload
    },
    setDisplaySortablePinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortablePinnedApps = payload
    },
    setDisplaySortableTemporaryPinnedApps: (state: AppState, { payload }: PayloadAction<boolean>) => {
      state.displaySortableTemporaryPinnedApps = payload
    },
    // AppsSearchState
    setAppsSearchQuery: (state: AppState, { payload }: PayloadAction<string | undefined>) => {
      state.search.query = payload
    },
    setAppsSearchResult: (state: AppState, { payload }: PayloadAction<AppDetails[]>) => {
      state.search.result = payload
    },
    resetAppsSearchState: (state: AppState) => {
      state.search.query = undefined
      state.search.result = []
    },
  },
})

export const {
  setDisplayAllApps,
  setDisplaySettings,
  setMenuAppDetails,
  setDisplayAppMenu,
  setDisplaySortableFavoriteApps,
  setDisplaySortablePinnedApps,
  setDisplaySortableTemporaryPinnedApps,
  setAppsSearchQuery,
  setAppsSearchResult,
  resetAppsSearchState,
} = appStateSlice.actions

export const toogleAllApps = createAction('appState/toogleAllApps')
export const appLaunch = createAction<{ renderedIn: RenderedIn; appDetails: AppDetails }>('appState/appLaunch')
export const appLaunchFromSearch = createAction('appState/appLaunchFromSearch')
export const sortFavoriteApps = createAction('appState/sortFavoriteApps')
export const sortPinnedApps = createAction('appState/sortPinnedApps')
export const sortTemporaryPinnedApps = createAction('appState/sortTemporaryPinnedApps')

export const selectAppState = (state: RootState) => state.appState
export const selectDisplayAllApps = (state: RootState) => state.appState.displayAllApps
export const selectDisplaySettings = (state: RootState) => state.appState.displaySettings
export const selectMenuAppDetails = (state: RootState) => state.appState.menuAppDetails
export const selectDisplayAppMenu = (state: RootState) => state.appState.displayAppMenu
export const selectDisplaySortableFavoriteApps = (state: RootState) => state.appState.displaySortableFavoriteApps
export const selectDisplaySortablePinnedApps = (state: RootState) => state.appState.displaySortablePinnedApps
export const selectDisplaySortableTemporaryPinnedApps = (state: RootState) => state.appState.displaySortableTemporaryPinnedApps
export const selectAppsSearchQuery = (state: RootState) => state.appState.search.query
export const selectAppsSearchResult = (state: RootState) => state.appState.search.result

export default appStateSlice.reducer
