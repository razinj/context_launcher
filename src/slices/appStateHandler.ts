import { PayloadAction } from '@reduxjs/toolkit'
import { call, put, select, takeLatest } from 'redux-saga/effects'
import { AppDetails } from '../models/app-details'
import { FavoriteApp } from '../models/favorite-app'
import { PinnedApp } from '../models/pinned-app'
import { RenderedIn } from '../models/rendered-in'
import { launchApp } from '../utils/apps-module'
import { dismissKeyboard } from '../utils/keyboard'
import {
  appLaunch,
  appLaunchFromSearch,
  resetAppsSearchState,
  selectAppsSearchResult,
  selectDisplayAllApps,
  setDisplayAllApps,
  setDisplaySettings,
  setDisplaySortableFavoriteApps,
  setDisplaySortablePinnedApps,
  setDisplaySortableTemporaryPinnedApps,
  setMenuAppDetails,
  sortFavoriteApps,
  sortPinnedApps,
  sortTemporaryPinnedApps,
  toogleAllApps,
} from './appState'
import { selectFavoriteApps } from './favoriteApps'
import { selectPinnedApps } from './pinnedApps'
import { addRecentApp } from './recentApps'

// We should only add a launched app if it was clicked (rendered) in one of the following places;
const ADD_TO_RECENT_APPS_RENDERED_IN_VALUES = [RenderedIn.ALL_APPS, RenderedIn.FILTERED_APPS]

function* toggleAllAppsHandler() {
  const allAppsDisplayed: boolean = yield select(selectDisplayAllApps)
  yield put(setDisplayAllApps(!allAppsDisplayed))
}

function* appLaunchHandler(action: PayloadAction<{ renderedIn: RenderedIn; appDetails: AppDetails }>) {
  yield call(dismissKeyboard)
  yield put(resetAppsSearchState())
  yield put(setDisplayAllApps(false))
  yield put(setMenuAppDetails(undefined))
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableFavoriteApps(false))

  const { appDetails, renderedIn } = action.payload

  if (!ADD_TO_RECENT_APPS_RENDERED_IN_VALUES.includes(renderedIn)) {
    return
  }

  // Skip adding to recent apps if permanently pinned/favourited but launched from different lists/views
  const pinnedApps: PinnedApp[] = yield select(selectPinnedApps)
  const favoriteApps: FavoriteApp[] = yield select(selectFavoriteApps)
  // TODO: Include the temporarily pinned apps but only when they're rendered/displayed.

  const pinnedAppsPackages: string[] = pinnedApps.map(({ packageName }: PinnedApp) => packageName)
  const favoriteAppsPackages: string[] = favoriteApps.map(({ packageName }: FavoriteApp) => packageName)
  const allApps = new Set([...pinnedAppsPackages, ...favoriteAppsPackages])

  if (allApps.has(appDetails.packageName)) {
    return
  }

  // Otherwise add to recent apps
  yield put(addRecentApp(appDetails))
}

function* sortFavoriteAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableFavoriteApps(true))
}

function* sortPinnedAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortablePinnedApps(true))
}

function* sortTemporaryPinnedAppsHandler() {
  yield put(setDisplaySettings(false))
  yield put(setDisplaySortableTemporaryPinnedApps(true))
}

function* appLaunchFromSearchHandler() {
  const searchResult: AppDetails[] = yield select(selectAppsSearchResult)
  if (searchResult.length === 0) {
    return
  }

  const appDetails = searchResult[0]

  yield call(launchApp, appDetails.packageName)
  yield put(appLaunch({ renderedIn: RenderedIn.FILTERED_APPS, appDetails }))
}

export function* appStateSaga() {
  yield takeLatest(toogleAllApps.type, toggleAllAppsHandler)
  yield takeLatest(appLaunch.type, appLaunchHandler)
  yield takeLatest(sortFavoriteApps.type, sortFavoriteAppsHandler)
  yield takeLatest(sortPinnedApps.type, sortPinnedAppsHandler)
  yield takeLatest(sortTemporaryPinnedApps.type, sortTemporaryPinnedAppsHandler)
  yield takeLatest(appLaunchFromSearch.type, appLaunchFromSearchHandler)
}
