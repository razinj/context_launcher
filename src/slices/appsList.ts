import { createAction, createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { AppDetails } from '../models/app-details'
import { RootState } from '../store'
import { getAppsLetterIndex } from '../utils/alphabet-list'
import { displayToast } from '../utils/toast'

export interface AppsListState {
  list: AppDetails[]
}

const initialState: AppsListState = {
  list: [],
}

export const appsListSlice = createSlice({
  name: 'appsList',
  initialState,
  reducers: {
    setAppsList: (state: AppsListState, { payload }: PayloadAction<AppDetails[]>) => {
      // console.log('setAppsList called');
      displayToast(`apps found : ${payload.length}`)

      state.list = payload.sort((appOne: AppDetails, appTwo: AppDetails) => appOne.name.localeCompare(appTwo.name))
    },
  },
})

export const { setAppsList } = appsListSlice.actions

export const getAppsListAction = createAction('appsList/getAppsListAction')
export const appRemovedAction = createAction<string>('appsList/appRemovedAction')
export const revalidateAppsLists = createAction('appsList/revalidateAppsLists')

const selectAppsList = (state: RootState) => state.appsList.list

export const selectAppsListMemoized = createSelector(selectAppsList, (list: AppDetails[]) => list)
export const selectAppsLetterListMemoized = createSelector(selectAppsList, (list: AppDetails[]) =>
  getAppsLetterIndex(list)
)

export default appsListSlice.reducer
