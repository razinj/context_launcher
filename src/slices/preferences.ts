// Redux
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
// Models
import { RootState } from '../store'

export interface PreferencesState {
  displayRecentApps: boolean
  displayFavoriteApps: boolean
  displayPinnedApps: boolean
  displayTemporaryPinnedApps: boolean
}

const initialState: PreferencesState = {
  displayRecentApps: true,
  displayFavoriteApps: true,
  displayPinnedApps: true,
  displayTemporaryPinnedApps: false,
}

export const preferencesSlice = createSlice({
  name: 'preferences',
  initialState,
  reducers: {
    displayRecentApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayRecentApps = payload
    },
    displayFavoriteApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayFavoriteApps = payload
    },
    displayPinnedApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayPinnedApps = payload
    },
    displayTemporaryPinnedApps: (state: PreferencesState, { payload }: PayloadAction<boolean>) => {
      state.displayTemporaryPinnedApps = payload
    },
    resetPreferences: (state: PreferencesState) => {
      state = { ...initialState }
    },
  },
})

export const {
  displayRecentApps,
  displayFavoriteApps,
  displayPinnedApps,
  displayTemporaryPinnedApps,
  resetPreferences,
} = preferencesSlice.actions

const selectDisplayRecentApps = (state: RootState) => state.preferences.displayRecentApps
const selectDisplayFavoriteApps = (state: RootState) => state.preferences.displayFavoriteApps
const selectDisplayPinnedApps = (state: RootState) => state.preferences.displayPinnedApps
const selectDisplayTemporaryPinnedApps = (state: RootState) => state.preferences.displayTemporaryPinnedApps

export const selectDisplayRecentAppsMemoized = createSelector(selectDisplayRecentApps, (value: boolean) => value)
export const selectDisplayFavoriteAppsMemoized = createSelector(selectDisplayFavoriteApps, (value: boolean) => value)
export const selectDisplayPinnedAppsMemoized = createSelector(selectDisplayPinnedApps, (value: boolean) => value)
export const selectDisplayTemporaryPinnedAppsMemoized = createSelector(
  selectDisplayTemporaryPinnedApps,
  (value: boolean) => value
)

export default preferencesSlice.reducer
