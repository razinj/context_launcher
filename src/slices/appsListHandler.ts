import { PayloadAction } from '@reduxjs/toolkit'
import { Channel, channel } from 'redux-saga'
import { call, fork, put, select, take, takeLatest } from 'redux-saga/effects'
import { AppDetails } from '../models/app-details'
import LauncherAppsModule from '../native-modules/LauncherAppsModule'
import { appRemovedAction, getAppsListAction, revalidateAppsLists, setAppsList } from './appsList'
import { removeFavoriteApp, removeFavoriteApps, selectFavoriteAppsMemoized } from './favoriteApps'
import {
  removePinnedApp,
  removePinnedApps,
  selectPinnedAppsMemoized,
  selectTemporaryPinnedAppsMemoized,
} from './pinnedApps'
import { removeRecentApp, removeRecentApps, selectRecentAppsMemoized } from './recentApps'

const notAvailable = 'NOT_AVAILABLE'

function* getAppsListActionHandler() {
  try {
    const applications: string = yield call(LauncherAppsModule.getApplications)
    const apps = JSON.parse(applications) as AppDetails[]
    yield put(setAppsList(apps))
  } catch (error: unknown) {
    yield put(setAppsList([]))
    console.error('Error getting apps list:', error)
  }
}

function* appRemovedActionHandler(action: PayloadAction<string>) {
  yield put(removeRecentApp(action.payload))
  const pinnedApp = { id: notAvailable, packageName: action.payload, name: notAvailable, icon: notAvailable }
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: true })) // TODO: Check why I send an "app" inseatd of "packageName" for removal
  yield put(removePinnedApp({ app: pinnedApp, isPermanent: false }))
  yield put(removeFavoriteApp(action.payload))
}

// Revalidate all apps' lists
function* x() {
  console.log('x called')

  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  yield put(setAppsList(apps))

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

// const upsertAppsListChannel = channel()

// function* watchAppsListUpsert() {
//   while (true) {
//     type x = typeof setAppsList
//     const action: x = yield take(upsertAppsListChannel)
//     yield put(action)
//   }
// }

function getApplicationsV2Async() {
  return new Promise((resolve, reject) => {
    LauncherAppsModule.getApplicationsV2((applications: string) => {
      resolve(applications)
    })
  })
}

function* y() {
  console.log('y called')

  const applications: string = yield call(getApplicationsV2Async)
  yield call(handleApplications, applications)

  const apps = JSON.parse(applications) as AppDetails[]

  const appsPackageNames = apps.map(({ packageName }: AppDetails) => packageName)
  // const appsPackageNames: string[] = []

  // Revalidate recent apps
  const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
  yield put(
    removeRecentApps(
      recentApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate favorite apps
  const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
  yield put(
    removeFavoriteApps(
      favoriteApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        )
    )
  )

  // Revalidate pinned apps
  const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: true,
      packageNames: pinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )

  // Revalidate temporary pinned apps
  const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
  yield put(
    removePinnedApps({
      isPermanent: false,
      packageNames: temporaryPinnedApps
        .map(({ packageName }: AppDetails) => packageName)
        .filter(
          (packageName: string) => !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
        ),
    })
  )
}

function* handleApplications(applications: string) {
  console.log('handleApplications called')

  const apps = JSON.parse(applications) as AppDetails[]
  // dispatch an action here
  yield put(setAppsList(apps))
}

function* childSaga(channel: Channel<any>) {
  const applications: string = yield call(LauncherAppsModule.getApplications)
  const apps = JSON.parse(applications) as AppDetails[]
  // yield put(setAppsList(apps))
  // const result = yield call(someAsyncFunction);
  yield put(channel, apps) // Emit the result through the channel
}

function* z() {
  console.log('z called')

  const communicationChannel: Channel<AppDetails[]> = yield call(channel)
  yield fork(childSaga, communicationChannel)

  while (true) {
    const result: AppDetails[] = yield take(communicationChannel)
    console.log(' result: ', result.length)

    const appsPackageNames = result.map(({ packageName }: AppDetails) => packageName)
    // const appsPackageNames: string[] = []

    // Revalidate recent apps
    const recentApps: AppDetails[] = yield select(selectRecentAppsMemoized)
    yield put(
      removeRecentApps(
        recentApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate favorite apps
    const favoriteApps: AppDetails[] = yield select(selectFavoriteAppsMemoized)
    yield put(
      removeFavoriteApps(
        favoriteApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          )
      )
    )

    // Revalidate pinned apps
    const pinnedApps: AppDetails[] = yield select(selectPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: true,
        packageNames: pinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Revalidate temporary pinned apps
    const temporaryPinnedApps: AppDetails[] = yield select(selectTemporaryPinnedAppsMemoized)
    yield put(
      removePinnedApps({
        isPermanent: false,
        packageNames: temporaryPinnedApps
          .map(({ packageName }: AppDetails) => packageName)
          .filter(
            (packageName: string) =>
              !appsPackageNames.some((appsPackageName: string) => appsPackageName === packageName)
          ),
      })
    )

    // Use the result returned by the child saga
    // Non-blocking logic here

    // ...
  }

  console.log('done')
}

export function* appsListSagas() {
  yield takeLatest(getAppsListAction.type, getAppsListActionHandler)
  yield takeLatest(appRemovedAction.type, appRemovedActionHandler)
  // yield takeLatest(revalidateAppsLists.type, x)
  // yield takeLatest(revalidateAppsLists.type, z)
  yield takeLatest(revalidateAppsLists.type, z)
}
