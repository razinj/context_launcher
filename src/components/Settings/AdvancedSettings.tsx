// React
import React from 'react'
// React Native
import { Pressable, PressableAndroidRippleConfig, StyleSheet, ToastAndroid, View } from 'react-native'
// Components
import SettingsItemLabel from './SettingsItemLabel'
// Redux
import { useDispatch } from 'react-redux'
import { setAppsList } from '../../slices/appsList'
import { clearRecentApps } from '../../slices/recentApps'
import { resetPreferences } from '../../slices/preferences'
import { clearFavoriteApps } from '../../slices/favoriteApps'
// Native modules
import AppsModule from '../../native-modules/AppsModule'
// Models
import { AppDetails } from '../../models/app-details'

const settingItemButtonRippleConfig: PressableAndroidRippleConfig = {
  borderless: false,
  foreground: true,
  color: '#e5e5e5',
}

const AdvancedSettings = () => {
  const dispatch = useDispatch()

  const reloadAllApps = () => {
    AppsModule.getApplications((applications: string) => {
      const apps = JSON.parse(applications) as AppDetails[]

      dispatch(setAppsList(apps))
      displayToast('All apps reloaded successfully!')
    })
  }

  const onClearRecentApps = () => {
    dispatch(clearRecentApps())
    displayToast('Recent apps cleared successfully!')
  }

  const onClearFavoriteApps = () => {
    dispatch(clearFavoriteApps())
    displayToast('Favorite apps cleared successfully!')
  }

  const onResetPreferences = () => {
    dispatch(resetPreferences())
    displayToast('Settings reset successfully!')
  }

  const displayToast = (message: string, duration: number = ToastAndroid.LONG) => {
    ToastAndroid.show(message, duration)
  }

  return (
    <>
      {/* Reload all apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-reload-all-apps-button'
          onPress={reloadAllApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Reload all apps' />
        </Pressable>
      </View>

      {/* Clear recent apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-clear-recent-apps-button'
          onPress={onClearRecentApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Clear recent apps' />
        </Pressable>
      </View>

      {/* Clear favorite apps */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-clear-favorite-apps-button'
          onPress={onClearFavoriteApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Clear favorite apps' />
        </Pressable>
      </View>

      {/* Reset preferences */}
      <View style={styles.itemContainer}>
        <Pressable
          testID='advanced-settings-reset-preferences-button'
          onPress={onResetPreferences}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={styles.buttonItemPressable}>
          <SettingsItemLabel title='Reset settings' />
        </Pressable>
      </View>
    </>
  )
}

const styles = StyleSheet.create({
  itemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    borderBottomWidth: 1,
    borderBottomColor: '#e5e5e5',
  },
  buttonItemPressable: {
    flex: 1,
  },
})

export default AdvancedSettings
