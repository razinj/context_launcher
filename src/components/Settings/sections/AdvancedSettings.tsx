// React
import React from 'react'
// React Native
import { Pressable, View } from 'react-native'
// Components
import SettingsItemLabel from '../shared/SettingsItemLabel'
// Redux
import { useDispatch } from 'react-redux'
import { setAppsList } from '../../../slices/appsList'
import { resetPreferences } from '../../../slices/preferences'
// Native modules
import AppsModule from '../../../native-modules/AppsModule'
// Utils
import { displayToast } from '../../../utils/toast'
// Models
import { AppDetails } from '../../../models/app-details'
// Constants
import { settingItemButtonRippleConfig, settingItemWrapperStyle, settingsPressableItemStyle } from '../shared/values'

const AdvancedSettings = () => {
  const dispatch = useDispatch()

  const reloadAllApps = () => {
    AppsModule.getApplications((applications: string) => {
      const apps = JSON.parse(applications) as AppDetails[]

      dispatch(setAppsList(apps))
      displayToast('All apps reloaded successfully!')
    })
  }

  const onResetPreferences = () => {
    dispatch(resetPreferences())
    displayToast('Settings reset successfully!')
  }

  return (
    <>
      <View style={settingItemWrapperStyle}>
        <Pressable
          testID='advanced-settings-reload-all-apps-button'
          onPress={reloadAllApps}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Reload all apps' />
        </Pressable>
      </View>

      <View style={settingItemWrapperStyle}>
        <Pressable
          testID='advanced-settings-reset-preferences-button'
          onPress={onResetPreferences}
          android_disableSound={true}
          android_ripple={settingItemButtonRippleConfig}
          style={settingsPressableItemStyle}>
          <SettingsItemLabel title='Reset settings' />
        </Pressable>
      </View>
    </>
  )
}

export default AdvancedSettings
