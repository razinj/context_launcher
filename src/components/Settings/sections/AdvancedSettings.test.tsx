import React from 'react'
import { fireEvent, screen } from '@testing-library/react-native'
import { initialStoreState } from '../../../../utils/test/data'
import { renderWithProvider } from '../../../../utils/test/utils'
import { ToastAndroid } from 'react-native'
import AppsModule from '../../../native-modules/AppsModule'
import AdvancedSettings from './AdvancedSettings'
import { PreferencesState } from '../../../slices/preferences'

describe('<AdvancedSettings /> Tests', () => {
  beforeEach(() => {
    jest.spyOn(ToastAndroid, 'show')
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should render correctly and match snapshot', () => {
    renderWithProvider(<AdvancedSettings />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('advanced-settings-reload-all-apps-button')).toBeOnTheScreen()
    expect(screen.getByTestId('advanced-settings-reset-preferences-button')).toBeOnTheScreen()
  })

  // TODO: Find a way to mock internal dispatched action
  it.todo('should call AppsModule method to get all applications and set apps list')

  it('should call AppsModule method to get all applications', () => {
    renderWithProvider(<AdvancedSettings />)

    const reloadAllAppsButton = screen.getByTestId('advanced-settings-reload-all-apps-button')

    expect(reloadAllAppsButton).toBeOnTheScreen()

    fireEvent.press(reloadAllAppsButton)

    expect(AppsModule.getApplications).toBeCalled()
  })

  it('should reset preferences apps when button is pressed', () => {
    const customInitialState = {
      ...initialStoreState,
      preferences: {
        displayRecentApps: false,
        displayFavoriteApps: true,
        displayPinnedApps: false,
        displayTemporaryPinnedApps: true,
      } as PreferencesState,
    }

    const { store } = renderWithProvider(<AdvancedSettings />, { preloadedState: customInitialState })

    const resetPreferencesButton = screen.getByTestId('advanced-settings-reset-preferences-button')

    expect(resetPreferencesButton).toBeOnTheScreen()

    fireEvent.press(resetPreferencesButton)

    expect(store.getState().preferences).toEqual({
      displayRecentApps: false,
      displayFavoriteApps: true,
      displayPinnedApps: false,
      displayTemporaryPinnedApps: true,
    })
    expect(ToastAndroid.show).toBeCalledWith('Settings reset successfully!', ToastAndroid.LONG)
  })
})
