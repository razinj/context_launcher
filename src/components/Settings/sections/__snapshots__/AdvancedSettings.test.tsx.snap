// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdvancedSettings /> Tests should render correctly and match snapshot 1`] = `
[
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
      testID="advanced-settings-reload-all-apps-button"
    >
      <View
        style={
          [
            {
              "padding": 10,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#808080",
              },
              undefined,
            ]
          }
        >
          Reload all apps
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
      testID="advanced-settings-reset-preferences-button"
    >
      <View
        style={
          [
            {
              "padding": 10,
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#808080",
              },
              undefined,
            ]
          }
        >
          Reset settings
        </Text>
      </View>
    </View>
  </View>,
]
`;
